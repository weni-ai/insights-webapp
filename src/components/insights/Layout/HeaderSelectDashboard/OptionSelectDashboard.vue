<template>
  <UnnnicDropdownItem
    data-testid="option-select-dashboard"
    class="option-select-dashboard"
    :class="{
      'option-select-dashboard--active':
        currentDashboard.uuid === dashboard.uuid,
    }"
    @click="setCurrentDashboard(dashboard)"
  >
    {{ dashboard.name }}
    <UnnnicIcon
      data-testid="star-icon"
      class="option-select-dashboard__star-icon"
      :class="{
        'option-select-dashboard__star-icon--selected': isDefaultDashboard,
      }"
      icon="star_rate"
      scheme="neutral-clean"
      clickable
      :filled="isDefaultDashboard || starHovered"
      @mouseenter="setStarHovered(true)"
      @mouseleave="setStarHovered(false)"
      @click.stop="handleSetDefaultDashboard"
    />
  </UnnnicDropdownItem>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import Unnnic from '@weni/unnnic-system';

export default {
  name: 'OptionSelectDashboard',

  props: {
    dashboard: {
      type: Object,
      default: () => {},
    },
  },

  data() {
    return {
      dashboardHovered: '',
      starHovered: false,
    };
  },

  computed: {
    ...mapState({
      currentDashboard: (state) => state.dashboards.currentDashboard,
    }),
    ...mapGetters({
      dashboardDefault: 'dashboards/dashboardDefault',
    }),

    isDefaultDashboard() {
      return this.dashboardDefault?.uuid === this.dashboard.uuid;
    },
  },

  methods: {
    ...mapActions({
      setCurrentDashboard: 'dashboards/setCurrentDashboard',
      setDefaultDashboard: 'dashboards/setDefaultDashboard',
    }),

    setStarHovered(boolean) {
      this.starHovered = boolean;
    },

    async handleSetDefaultDashboard() {
      const { dashboard } = this;
      if (dashboard.uuid === this.dashboardDefault.uuid) return;

      try {
        await this.setDefaultDashboard(dashboard.uuid);
        this.callSetDashboardAlert('success');
      } catch (error) {
        console.error(error);
        this.callSetDashboardAlert('error');
      }
    },

    callSetDashboardAlert(type) {
      const { dashboard } = this;

      if (!['success', 'error'].includes(type)) {
        throw new Error(
          'Error calling the alert when setting the default dashboard. This type does not exist.',
        );
      }

      Unnnic.unnnicCallAlert({
        props: {
          text: this.$t(`insights_header.set_default_dashboard_${type}`, {
            dashboard: dashboard.name,
          }),
          type,
        },
        seconds: 5,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
.option-select-dashboard {
  &.unnnic-dropdown-item::before {
    content: none;
  }

  border-radius: $unnnic-border-radius-sm;

  padding: $unnnic-spacing-xs;

  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: $unnnic-spacing-nano;

  color: $unnnic-color-neutral-darkest;
  font-family: $unnnic-font-family-secondary;
  font-size: $unnnic-font-size-body-gt;

  &--active {
    background-color: $unnnic-color-neutral-lightest;
    font-weight: $unnnic-font-weight-bold;
  }

  .option-select-dashboard__star-icon:not(
      .option-select-dashboard__star-icon--selected
    ):hover {
    color: $unnnic-color-weni-500;
  }

  .option-select-dashboard__star-icon--selected {
    color: $unnnic-color-weni-600;
  }
}
</style>
