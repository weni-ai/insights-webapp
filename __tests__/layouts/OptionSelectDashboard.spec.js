import { beforeEach, describe, expect, it, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';

import OptionSelectDashboard from '@/components/insights/Layout/HeaderSelectDashboard/OptionSelectDashboard.vue';
import Unnnic from '@weni/unnnic-system';

const dashboard1 = { name: 'Dashboard 1', uuid: '1' };
const dashboard2 = { name: 'Dashboard 2', uuid: '2' };

describe('OptionSelectDashboard', () => {
  let wrapper;
  const store = createStore({
    modules: {
      dashboards: {
        namespaced: true,
        state: {
          dashboards: [dashboard1, dashboard2],
          currentDashboard: dashboard1,
        },
        getters: {
          dashboardDefault: () => dashboard1,
        },
        actions: {
          setCurrentDashboard({ state }, dashboard) {
            state.currentDashboard = dashboard;
          },
          setDefaultDashboard: vi.fn(() => Promise.resolve()),
        },
      },
      config: {
        namespaced: true,
        state: {
          enableCreateCustomDashboards: false,
        },
      },
    },
  });

  const createWrapper = (props) => {
    return mount(OptionSelectDashboard, {
      props,
      global: {
        plugins: [store],
        stubs: {
          UnnnicDropdownItem: Unnnic.unnnicDropdownItem,
          UnnnicIcon: true,
        },
        mocks: {
          $t: (msg) => msg,
        },
      },
    });
  };

  let starIcon;
  beforeEach(() => {
    wrapper = createWrapper({ dashboard: dashboard1 });
    starIcon = wrapper.findComponent('[data-testid=star-icon]');
  });

  describe('Rendering and Interactions', () => {
    it('Should display the name of the dashboard prop', () => {
      const optionSelectDashboard = wrapper.findComponent(
        '[data-testid=option-select-dashboard]',
      );

      expect(optionSelectDashboard.exists()).toBe(true);
      expect(optionSelectDashboard.text()).toContain('Dashboard 1');
    });

    it('Should set the current dashboard when an option is clicked', async () => {
      wrapper = createWrapper({
        dashboard: dashboard2,
      });

      const optionSelectDashboard = wrapper.findComponent(
        '[data-testid=option-select-dashboard]',
      );
      await optionSelectDashboard.trigger('click');

      expect(store.state.dashboards.currentDashboard.uuid).toBe('2');
    });

    it('Should have a star_rate icon', () => {
      expect(starIcon.exists()).toBe(true);
      expect(starIcon.props('icon')).toBe('star_rate');
    });

    it('Should fill the star_rate icon if the dashboard is default or on hover', async () => {
      expect(starIcon.classes()).toContain(
        'option-select-dashboard__star-icon--selected',
      );
      expect(starIcon.props('filled')).toBe(true);

      await wrapper.setProps({ dashboard: dashboard2 });

      await starIcon.trigger('mouseenter');
      expect(starIcon.props('filled')).toBe(true);

      await starIcon.trigger('mouseleave');
      expect(starIcon.props('filled')).toBe(false);
    });
  });

  describe('Default Dashboard Functionality', () => {
    it('Should set default dashboard when star icon and dashboard is different from default', async () => {
      const spySetDefaultDashboard = vi.spyOn(
        wrapper.vm,
        'setDefaultDashboard',
      );

      await starIcon.trigger('click');

      expect(spySetDefaultDashboard).not.toHaveBeenCalled();

      await wrapper.setProps({ dashboard: dashboard2 });
      await starIcon.trigger('click');

      expect(spySetDefaultDashboard).toHaveBeenCalled();
    });

    it('Should show a success alert if the default dashboard is set', async () => {
      await wrapper.setProps({ dashboard: dashboard2 });
      const spyUnnnicAlert = vi
        .spyOn(Unnnic, 'unnnicCallAlert')
        .mockImplementation(() => {});
      const createAlertMsg = (type) =>
        wrapper.vm.$t(`insights_header.set_default_dashboard_${type}`, {
          dashboard: dashboard2.name,
        });

      wrapper.vm.setDefaultDashboard = vi.fn(() => Promise.reject());
      await starIcon.trigger('click');
      expect(spyUnnnicAlert).toHaveBeenCalledWith({
        props: {
          text: createAlertMsg('error'),
          type: 'error',
        },
        seconds: 5,
      });

      wrapper.vm.setDefaultDashboard = vi.fn(() => Promise.resolve());
      await starIcon.trigger('click');
      expect(spyUnnnicAlert).toHaveBeenCalledWith({
        props: {
          text: createAlertMsg('success'),
          type: 'success',
        },
        seconds: 5,
      });
    });
  });

  it('Should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
