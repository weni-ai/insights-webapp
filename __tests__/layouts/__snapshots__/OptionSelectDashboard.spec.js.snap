// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OptionSelectDashboard > Should match the snapshot 1`] = `
<a
  class="unnnic-dropdown-item unnnic--clickable option-select-dashboard"
  data-testid="option-select-dashboard"
  data-v-000aa0c0=""
>
  
  Dashboard 1 
  <span
    class="material-symbols-rounded unnnic-icon-scheme--neutral-clean unnnic-icon-size--md unnnic--clickable material-symbols-rounded--filled option-select-dashboard__star-icon option-select-dashboard__star-icon--selected"
    data-testid="star-icon"
    data-v-000aa0c0=""
    data-v-76ceda5d=""
    translate="no"
  >
    star_rate
  </span>
  
</a>
`;
